src/discordcr/voice.cr

```diff
@@ -0,0 +1,268 @@
+      LOGGER.debug("VWS packet received: #{packet} #{packet.data.to_s}")
+
+      case packet.opcode
+      when OP_READY
+        payload = VWS::ReadyPayload.from_json(packet.data)
+        handle_ready(payload)
+      when OP_SESSION_DESCRIPTION
+        payload = VWS::SessionDescriptionPayload.from_json(packet.data)
+        handle_session_description(payload)
+      when OP_HELLO
+        payload = VWS::HelloPayload.from_json(packet.data)
+        handle_hello(payload)
+      end
+    end
+
+    private def on_close(message : String)
+      LOGGER.warn("VWS closed with message: #{message}")
+      nil
+    end
+
+    private def handle_ready(payload : VWS::ReadyPayload)
+      udp_connect(payload.port.to_u32, payload.ssrc.to_u32)
+    end
+
+    private def udp_connect(port, ssrc)
+      @udp.connect(@endpoint, port, ssrc)
+      @udp.send_discovery
+      ip, port = @udp.receive_discovery_reply
+      send_select_protocol(UDP_PROTOCOL, ip, port, ENCRYPTED_MODE)
+    end
+
+    private def send_identify(server_id, user_id, session_id, token)
+      packet = VWS::IdentifyPacket.new(server_id, user_id, session_id, token)
+      @websocket.send(packet.to_json)
+    end
+
+    private def send_select_protocol(protocol, address, port, mode)
+      data = VWS::ProtocolData.new(address, port, mode)
+      packet = VWS::SelectProtocolPacket.new(protocol, data)
+      @websocket.send(packet.to_json)
+    end
+
+    private def handle_session_description(payload : VWS::SessionDescriptionPayload)
+      @udp.secret_key = payload.secret_key
+
+      # Once the secret key has been received, we are ready to send audio data.
+      # Notify the user of this
+      spawn { @ready_handler.try(&.call) }
+    end
+
+    private def handle_hello(payload : VWS::HelloPayload)
+      @heartbeat_interval = payload.heartbeat_interval
+      send_identify(@server_id, @user_id, @session_id, @token)
+    end
+  end
+
+  # Client for Discord's voice UDP protocol, on which the actual audio data is
+  # sent. There should be no reason to manually use this class: use
+  # `VoiceClient` instead which uses this class internally.
+  class VoiceUDP
+    @secret_key : Array(UInt8)?
+    property secret_key
+
+    def initialize
+      @socket = UDPSocket.new
+    end
+
+    def connect(endpoint : String, port : UInt32, ssrc : UInt32)
+      endpoint = endpoint[6..-1] if endpoint.starts_with? "wss://"
+      endpoint = endpoint.gsub(":80", "") # The endpoint may contain a port, we don't want that
+
+      @ssrc = ssrc
+      @socket.connect(endpoint, port)
+    end
+
+    # Sends a discovery packet to Discord, telling them that we want to know our
+    # IP so we can select the protocol on the VWS
+    def send_discovery
+      io = IO::Memory.new(70)
+
+      io.write_bytes(@ssrc.not_nil!, IO::ByteFormat::BigEndian) # [@ssrc].pack("N")
+      io.write(Bytes.new(66, 0_u8))                             # 66 null bytes for a total of 70 bytes
```

Comment:
Id use 70 - sizeof(Int32) here. It's a bit more self-explanitory. Also can we loose the ruby translations, its more than possible to see what this code does without looking at the docs (unlike ruby).